test/integration/imageimporter_test.go

```diff
@@ -183,6 +183,134 @@ func mockRegistryHandler(t *testing.T, requireAuth bool, count *int) http.Handle
 	})
 }
 
+func TestImageStreamImportFromV2Repository(t *testing.T) {
+	imageDigest := "sha256:815d06b56f4138afacd0009b8e3799fcdce79f0507bf8d0588e219b93ab6fd4d"
+	descriptors := map[string]int64{
+		"sha256:a3ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4": 3000,
+		"sha256:86e0e091d0da6bde2456dbb48306f3956bbeb2eae1b5b9a43045843f69fe4aaa": 200,
+		"sha256:b4ed95caeb02ffe68cdd9fd84406680ae93d633cb16422d00e8a7c22955b46d4": 10,
+	}
+
+	testutil.RequireEtcd(t)
+
+	// start regular HTTP servers
+	requireAuth := false
+	count := 0
+	countStat := 0
+
+	server := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		count++
+		t.Logf("%d got %s %s", count, r.Method, r.URL.Path)
+
+		w.Header().Set("Docker-Distribution-API-Version", "registry/2.0")
+		if requireAuth {
+			if len(r.Header.Get("Authorization")) == 0 {
+				w.Header().Set("WWW-Authenticate", "BASIC")
+				w.WriteHeader(http.StatusUnauthorized)
+				return
+			}
+		}
+
+		switch r.URL.Path {
+		case "/v2/":
+			w.Write([]byte(`{}`))
+		case "/v2/test/image/tags/list":
+			w.Write([]byte("{\"name\": \"test/image\", \"tags\": [\"testtag\"]}"))
+		case "/v2/test/image/manifests/testtag", "/v2/test/image/manifests/" + imageDigest:
+			if r.Method == "HEAD" {
+				w.Header().Set("Content-Length", fmt.Sprintf("%d", len(convertedManifest)))
+				w.Header().Set("Docker-Content-Digest", imageDigest)
+				w.WriteHeader(http.StatusOK)
+			} else {
+				w.Write([]byte(convertedManifest))
+			}
+		default:
+			if strings.HasPrefix(r.URL.Path, "/v2/test/image/blobs/") {
+				for dgst, size := range descriptors {
+					if r.URL.Path != "/v2/test/image/blobs/"+dgst {
+						continue
+					}
+					if r.Method == "HEAD" {
+						w.Header().Set("Content-Length", fmt.Sprintf("%d", size))
+						w.Header().Set("Docker-Content-Digest", dgst)
+						w.WriteHeader(http.StatusOK)
+						countStat++
+						return
+					}
+				}
+			}
+			t.Fatalf("unexpected request %s: %#v", r.URL.Path, r)
+		}
```

Comment:
There's a high potential for sharing most of this handler's code with the other tests.